### PR TITLE
Mr Solver precondition hints

### PR DESCRIPTION
**Pull request by eddywestbrook**
This PR adds support for "precondition hints" to Mr Solver. These hints are in the form of calls to the dummy function `precondHint` as the top level of the definition of a recursive function, and are used to communicate loop invariants to Mr Solver.

This PR also fixes a bug in the `smalledBitSetElem` function in the SAW core pretty-printer.

### CODE REVIEW

File: src/SAWScript/Prover/MRSolver/Monad.hs [hunk 1]
```diff
@@ -252,6 +246,8 @@ asEither _ = Nothing
 -- | A map from 'Term's to 'DataTypeAssump's over that term
 type DataTypeAssumps = HashMap Term DataTypeAssump
 
+-- FIXME HERE NOW: remove preconditions from MREnv
+
```
Comment (m-yac) [inline src/SAWScript/Prover/MRSolver/Monad.hs:250]: Was this resolved? If so this comment should probably be removed.

File: src/SAWScript/Prover/MRSolver/Monad.hs [hunk 2]
```diff
@@ -790,9 +799,9 @@ mrGetCoIndHyp :: FunName -> FunName -> MRM (Maybe CoIndHyp)
 mrGetCoIndHyp nm1 nm2 = Map.lookup (nm1, nm2) <$> mrCoIndHyps
 
 -- | Run a compuation under an additional co-inductive assumption
-withCoIndHypRaw :: CoIndHyp -> MRM a -> MRM a
-withCoIndHypRaw hyp m =
-  do debugPretty 1 ("withCoIndHyp" <+> ppInEmptyCtx hyp)
+withCoIndHyp :: CoIndHyp -> MRM a -> MRM a
+withCoIndHyp hyp m =
+  do debugPretty 2 ("withCoIndHypRaw" <+> ppInEmptyCtx hyp)
```
Comment (m-yac) [inline src/SAWScript/Prover/MRSolver/Monad.hs:804]: This should probably be kept as `"withCoIndHyp"` given that it's now the name of this function.

File: src/SAWScript/Prover/MRSolver/Monad.hs [hunk 3]
```diff
@@ -831,6 +861,28 @@ instantiateFunAssump fassump =
      rhs <- substTermLike 0 evars $ fassumpRHS fassump
      return (args, rhs)
 
+-- FIXME HERE NOW: delete this and remove the preconditions from the env
+-- | Look up the precondition associated with a function name, if there is one
+-- mrLookupPrecond :: FunName -> MRM (Maybe Term)
+-- mrLookupPrecond nm = Map.lookup nm <$> mrePreconditions <$> mriEnv <$> ask
+
```
Comment (m-yac) [inline src/SAWScript/Prover/MRSolver/Monad.hs:868]: As with the comment above, looks like this is done, so it should probably be removed?

File: src/SAWScript/Prover/MRSolver/Solver.hs [hunk 4]
```diff
@@ -345,31 +349,65 @@ handling the recursive ones
...
+  do let f1 = coIndHypLHSFun hyp
+         f2 = coIndHypRHSFun hyp
+         args1 = coIndHypLHS hyp
+         args2 = coIndHypRHS hyp
+     debugPretty 1 ("proveCoIndHyp" <+> ppInEmptyCtx hyp)
+     lhs <- fromMaybe (error "proveCoIndHyp") <$> mrFunBody f1 args1
+     rhs <- fromMaybe (error "proveCoIndHyp") <$> mrFunBody f2 args2
+     (pre1, pre2) <- applyCoIndHypPreconds hyp
+     pre <- liftSC2 scAnd pre1 pre2
+     (withOnlyUVars (coIndHypCtx hyp) $ withOnlyAssumption pre $
```
Comment (m-yac) [inline src/SAWScript/Prover/MRSolver/Solver.hs:399]: I feel like we loose some generality here by getting rid of all our other assumptions. Like, what if we have nested co-inductions and we need a precondition from the outer co-induction to prove something in the inner? Though we don't have any examples like that as the moment, so it's probably fine for now.

**Review by m-yac — APPROVED** [review]
Looks good to me! Just a few nitpicks, if you don't want to have to wait for the CI, I'm happy to just include the changes in my upcoming PR.

File: src/SAWScript/Prover/MRSolver/Monad.hs (hunk 2 above)
Comment (eddywestbrook) [reply src/SAWScript/Prover/MRSolver/Monad.hs:804]: Whoops, yes, it should be

File: src/SAWScript/Prover/MRSolver/Monad.hs (hunk 1 above)
Comment (eddywestbrook) [reply src/SAWScript/Prover/MRSolver/Monad.hs:250]: Whoops, yep, should be removed

File: src/SAWScript/Prover/MRSolver/Monad.hs (hunk 3 above)
Comment (eddywestbrook) [reply src/SAWScript/Prover/MRSolver/Monad.hs:868]: Yep

File: src/SAWScript/Prover/MRSolver/Solver.hs (hunk 4 above)
Comment (eddywestbrook) [reply src/SAWScript/Prover/MRSolver/Solver.hs:399]: Actually, the `withOnlyUVars` is going to throw away any existing assumptions anyway, so technically the `withOnlyAssumption` is superfluous, but it's helpful to have the "Only" there for documentation

**Pull request merged**